src/core/models.py

```diff
@@ -5,51 +5,87 @@
 """Collection of state objects for the Kafka relations, apps and units."""
 
 import logging
-from typing import MutableMapping
+from typing import Literal, MutableMapping
 
+from charms.data_platform_libs.v0.data_interfaces import Data, DataPeerData, DataPeerUnitData
 from charms.zookeeper.v0.client import QuorumLeaderNotFoundError, ZooKeeperManager
-from kazoo.exceptions import AuthFailedError, NoNodeError
+from kazoo.client import AuthFailedError, NoNodeError
 from ops.model import Application, Relation, Unit
-from tenacity import retry, retry_if_not_result, stop_after_attempt, wait_fixed
+from tenacity import retry
+from tenacity.retry import retry_if_not_result
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_fixed
+from typing_extensions import override
 
-from literals import INTERNAL_USERS, Substrate
+from literals import INTERNAL_USERS, SECRETS_APP
 
 logger = logging.getLogger(__name__)
 
+SUBSTRATES = Literal["vm", "k8s"]
 
-class StateBase:
-    """Base state object."""
+
+class RelationState:
+    """Relation state object."""
 
     def __init__(
-        self, relation: Relation | None, component: Unit | Application, substrate: Substrate
+        self,
+        relation: Relation | None,
+        data_interface: Data,
+        component: Unit | Application | None,
+        substrate: SUBSTRATES,
     ):
         self.relation = relation
+        self.data_interface = data_interface
         self.component = component
         self.substrate = substrate
+        self.relation_data = self.data_interface.as_dict(self.relation.id) if self.relation else {}
```

Comment:
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

WE MUSTN'T DO THIIS

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Imagine, that the developer (none of us of course ;-) ) does this.

```
# Imagine something like `peer_relation`... not there at this moment but 2 moments after
somestate = RelationState(relation=currently_not_but_later_existent_relation_property)

assert somestate.relation_data == {}

# THESE MODIFICATONS TAKE NO IMPACT
# WE ARE OPERATING ON A LCCAL DICT, NOT ON RELATION DATA
somestate['x'] = 'y'

# Even worse...
# Then the relation comes up
somestate['really_goes_on_the_relation'] = 'blah'
```

Now give the complex execution flows we have (x is calling y is calling z, etc.) it can get NASTY to follow with "fluctuating" relations, when do we write on a fake local dict, and when do we really modify relation data.

*******WE MUSTN'T DO THIS********